katsdpcontroller/product_config.py

```diff
@@ -31,63 +82,1196 @@ class _Sensor(ABC):
+                 adc_sample_rate: float,
+                 centre_frequency: float) -> None:
+        self.antenna_objects = list(antennas)
+        ratio = adc_sample_rate / (2 * bandwidth)
+        if abs(ratio - round(ratio)) > 1e-6:
+            raise ValueError('ADC Nyquist frequency is not a multiple of bandwidth')
+        n_samples_between_spectra = round(n_channels * adc_sample_rate // bandwidth)
+        super().__init__(
+            name, src_streams,
+            antennas=[antenna.name for antenna in self.antenna_objects],
+            band=band,
+            n_channels=n_channels,
+            bandwidth=bandwidth,
+            centre_frequency=centre_frequency,
+            adc_sample_rate=adc_sample_rate,
+            n_samples_between_spectra=n_samples_between_spectra
+        )
+
+    @classmethod
+    def from_config(cls,
+                    options: Options,
+                    name: str,
+                    config: Mapping[str, Any],
+                    src_streams: Sequence['Stream'],
+                    sensors: Mapping[str, Any]) -> 'SimAntennaChannelisedVoltageStream':
+        antennas = []
+        for desc in config['antennas']:
+            try:
+                antennas.append(katpoint.Antenna(desc))
+            except Exception as exc:
+                # katpoint can throw all kinds of exceptions
+                raise ValueError(f'Invalid antenna description {desc!r}: {exc}') from exc
+        return cls(
+            name, src_streams,
+            antennas=antennas,
+            band=config['band'],
+            n_channels=config['n_chans'],
+            bandwidth=config['bandwidth'],
+            adc_sample_rate=config['adc_sample_rate'],
+            centre_frequency=config['centre_frequency']
+        )
+
+
+class CbfPerChannelStream(Stream):
+    """Base for tied-array-channelised-voltage and baseline-correlation-products streams.
+
+    It applies to both simulated and real streams.
+    """
+
+    def __init__(self, name: str, src_streams: Sequence[Stream], *,
+                 n_endpoints: int,
+                 n_channels_per_substream: int,
+                 bits_per_sample: int) -> None:
+        super().__init__(name, src_streams)
+        self.n_endpoints = n_endpoints
+        self.n_channels_per_substream = n_channels_per_substream
+        self.bits_per_sample = bits_per_sample
+
+        if self.n_channels % self.n_endpoints != 0:
+            raise ValueError(
+                f'n_channels ({self.n_channels}) is not '
+                f'a multiple of endpoints ({self.n_endpoints})')
+        if self.n_channels_per_endpoint % self.n_channels_per_substream != 0:
+            raise ValueError(
+                f'channels per endpoint ({self.n_channels_per_endpoint}) '
+                f'is not a multiple of channels per substream ({self.n_channels_per_substream})'
+            )
+
+    @property
+    def antenna_channelised_voltage(self) -> 'AntennaChannelisedVoltageStreamBase':
```

Comment:
If I do that then any function that modifies `src_streams` will also need to remember to modify `antenna_channelised_voltage` (and if it's done in code that is working on the base class, it will need separate cases for each subclass). While I don't believe there is any such code at the moment, I prefer not to duplicate information for this reason.